content/blog/2022/05/verifying-image-signatures.md

```diff
@@ -0,0 +1,141 @@
+authors:
+- Raul Cabello Martin
+date: 2022-05-02
+---
+
+After these last releases Kubewarden now has support for verifying the integrity and authenticity of artifacts within
+Kubewarden using the Sigstore project. In this post, we shall focus on verifying container image signatures using the
+new [verify-image-signatures](https://github.com/kubewarden/verify-image-signatures) policy.
+
+To learn more about how Sigstore works, take a look at our
+previous [post](https://www.kubewarden.io/blog/2022/04/securing-kubewarden-policies/)
+
+## Verify Image Signatures Policy
+
+This policy validates Pods by checking their container images for signatures (that is, containers, init containers and 
+ephemeral containers in the pod).
+
+The policy can inspect all the container images defined inside of a Pod or it can just analyze the ones that
+are matching a pattern provided by the user.
+
+Container image tags are mutable, they can be changed to point to a completely different content. That's why it's a
+good security practice to reference container images by their immutable checksum.
+
+This policy can rewrite the image definitions that are using a tag to instead reference the image by its checksum.
+
+The policy will:
+  * Ensure the image referenced by a tag is satisfying the signature requested by the operator
+  * Extract the immutable reference of the image from the signatures
+  * Rewrite the image reference to be in the form `<image ref>@sha256:<digest>`
+This mutation can be disabled in the [settings](https://github.com/kubewarden/verify-image-signatures#settings).
+
+## Let's see it in action!
+
+For this example, a Kubernetes cluster with Kubewarden already installed is required. The installation process is
+described in the [quick start guide](https://docs.kubewarden.io/quick-start.html).
+
+We need an image with a signature that we can verify. You can use [cosign](https://github.com/sigstore/cosign) to sign
+your images. For this example we'll use the image
+`ghcr.io/viccuad/app-example:v0.1.0` that was signed using keyless verification. 
+
+Obtain the issuer and subject using cosign.
+
+```
+COSIGN_EXPERIMENTAL=1 cosign verify ghcr.io/viccuad/app-example:v0.1.0
+```
+
+```
+...
+"Issuer": "https://github.com/login/oauth",
+"Subject": "me@viccuad.me"
+...
+```
+
+Let's create a cluster-wide policy that will verify all images, and
+let's use the issuer and subject for verification:
+
+```
+kubectl apply -f - <<EOF
+apiVersion: policies.kubewarden.io/v1alpha2
+kind: ClusterAdmissionPolicy
+metadata:
+  name: verify-image-signatures
+spec:
+  module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.1.4
+  rules:
+  - apiGroups: [""]
+    apiVersions: ["v1"]
+    resources: ["pods"]
+    operations:
+    - CREATE
+    - UPDATE
+  mutating: true
+  settings:
+    signatures:
+      - image: "*"
+        keyless:
+          - issuer: "https://github.com/login/oauth"
+            subject: "me@viccuad.me"
```

Comment:
mhmhm I would like to remove this email, I already receive enough spam from mailing lists :sweat_smile: .

The image already has 2 signatures (see `cosign verify`), a manual one specifically for the presentation (the one with the email), and one from the workflow:
```
"Issuer":"https://token.actions.githubusercontent.com"
"Subject":"https://github.com/viccuad/app-example/.github/workflows/ci.yml@refs/tags/v0.1.0"
```

We can use this one for the post without any more changes.

Reply:
many thanks!